File: algorithms/cpp/stringCompression/stringCompression.cpp
// Source: https://leetcode.com/problems/string-compression/
// Author: Miao Zhang
// Date:   2021-02-12

class Solution {
public:
    int compress(vector<char>& chars) {
        int n = chars.size();
        int j = 0;
        for (int i = 1; i <=n; i++) {
            int cnt = 1;
            while (i < n && chars[i] == chars[i - 1]) {
                i++;
                cnt++;
            }
            chars[j++] = chars[i - 1];
            if (cnt == 1) continue;
            for (char c: to_string(cnt)) {
                chars[j++] = c;
            }
        }
        return j;
    }
};

File: algorithms/python/stringCompression/stringCompression.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/string-compression/
# Author: Miao Zhang
# Date:   2021-02-12

class Solution:
    def compress(self, chars: List[str]) -> int:
        n = len(chars)
        j = 0
        i = 1
        while i <= n:
            cnt = 1
            while i < n and chars[i] == chars[i - 1]:
                i += 1
                cnt += 1
            chars[j] = chars[i - 1]
            j += 1
            i += 1
            if cnt == 1: continue
            for c in str(cnt):
                chars[j] = c
                j += 1
        return j
